backends/gstreamer/src/webrtc.rs

```diff
@@ -146,14 +194,128 @@ impl GStreamerWebRtcController {
+            }
+        }
+        self.remote_mline_info.clear();
+        for i in 0..sdp.medias_len() {
+            let mut caps = gst::Caps::new_empty();
+            let caps_mut = caps.get_mut().unwrap();
+            let media = get_media(&sdp, i).unwrap();
+            for format in 0..media.formats_len() {
+                let pt = media.get_format(format).unwrap().parse().unwrap();
+                caps_mut.append(media.get_caps_from_media(pt).unwrap());
+                self.pt_counter = cmp::max(self.pt_counter, pt + 1);
+            }
+            for cap in 0..caps_mut.get_size() {
+                // the caps are application/x-unknown by default, which will fail
+                // to intersect
+                //
+                // see https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/blob/ba62917fbfd98ea76d4e066a6f18b4a14b847362/ext/webrtc/gstwebrtcbin.c#L2521
+                caps_mut.get_mut_structure(cap).unwrap().set_name("application/x-rtp")
+            }
+            self.remote_mline_info.push(MLineInfo {
+                caps: caps,
+                // XXXManishearth in the (yet unsupported) case of dynamic stream addition and renegotiation
+                // this will need to be checked against the current set of streams
+                is_used: false,
+                // XXXManishearth ideally, we keep track of all payloads and have the capability of picking
+                // the appropriate decoder. For this, a bunch of the streams code will have to be moved into
+                // a webrtc-specific abstraction.
+                payload: media.get_format(0).unwrap().parse().unwrap(),
+            });
+        }
+    }
+
+    /// Streams need to be linked to the correct pads, so we buffer them up until we know enough
+    /// to do this.
+    ///
+    /// When we get a remote offer, we store the relevant m-line information so that we can
+    /// pick the correct sink pad and payload. Shortly after we look for any pending streams
+    /// and connect them to available compatible m-lines using link_stream.
+    ///
+    /// When we create an offer, we're controlling the pad order, so we set request_new_pads
+    /// to true and forcefully link all pending streams before generating the offer.
+    fn link_stream(&mut self, mut boxed_stream: Box<MediaStream>, request_new_pads: bool) {
+        let stream = boxed_stream
+            .as_mut_any()
+            .downcast_mut::<GStreamerMediaStream>()
+            .unwrap();
+        let caps = stream.caps();
+        let webrtc = self.webrtc.as_ref().unwrap();
+        let idx = self.remote_mline_info.iter().enumerate()
+            .filter(|(_, x)| !x.is_used)
+            .find(|(_, x)| x.caps.can_intersect(&caps))
+            .map(|x| x.0);
+        let element = stream.src_element();
+
+        if let Some(idx) = idx {
+            if idx >= self.request_pad_counter {
```

Comment:
Is there any risk that we could call link_stream with !request_new_pads but end up running this code that requests pads? Would that be bad? I find the logic here a little bit obtuse.

Reply:
Yes, this code requesting new pads is fine, `request_new_pads` is about requesting new pads when there is no remote m-line: in that case we normally want to just buffer it up for later, except for when we're making an offer, at which point we want to flush all buffered streams.

The behavior of this function with `!request_new_pads` is "request pads only if they have been set up by the remote already", which is too long to fit in a variable. I can document this more explicitly.